### ISSUE TITLE
Braninstorming

### ISSUE THREAD

**Issue by Hube2** [Hube2/acf-fast-save#1]
So here's the problem. You can't just use and insert query to update values. If rows in the meta tables had to be unique then you could update or insert if it doesn't exist. But do to the fact that there is not uniqueness to a meta value this is why WP does at least two queries for every insert or update. First it needs to check to see if there is a value already, then it performs and insert or update depending on the result. WP then does another query to update the cache for a given meta name, which is completely unnecessary when saving a post since the next step is to be redirected.

The only way to speed this up would be to delete all of the acf fields by meta_name for a post and then to insert all of the new values. This reduces the number of queries to 2. You would also need to do another query to delete all repeater and flex field sub fields and this would need to be mulitple "LIKE" deletes.  So this gives us a total of three multiple statement queries rather than 

I'm not too sure about clone fields, I'll need to check on how these are saved in the database.

**Comment by Hube2** [comment]
There could be a problem when deleting all sub field values. I've run into this when creating fields that are readonly for editors or other specific types of users because the field gets deleted but not updated because the person cannot edit the field. So the only fields that you can delete are the ones where values have been submitted.

**Comment by Hube2** [comment]
I've been digging through the code in ACF and I've discovered something else that slows things down. Revisions. If revisions are supported by a post type then this will more than double the number of queries that need to be performed. Revisions is not something that I thought about. Before saving the values ACF gets all the existing values and copies them to the most recent revision and every field goes through the same update process for the revision.

This means that the save time on posts can immediately re reduced by at least 50% by disabling revisions. Which also explains why the same field group on an options page takes significantly less time to update than on a post page.

Revisions are not supported for term meta, user meta or options. I personally don't feel that revisions are that important. They waist time and space in the database for something that is rarely ever used. In fact, I know from personal experience that a posts with a huge number of ACF field crash a site when someone attempts to view/restore revisions, making them rather useless.

On another note. I've had to reduce the number of revisions stored on one of my sites for all posts with a specific field group. The reason is that for "Editors" just trying to edit one of these posts crashed the admin. It did not crash if the user was an admin. There must be something significantly different about how WP loads the edit post page for editors as apposed to admins.

So I guess that leaves the question, should ACF even be supporting revisions at all.

**Comment by jakefinley** [comment]
I can say from experience that my clients (and boss) appreciate the revisions feature. In particular since ACF typically contains all the content of the page when using it as a page builder. Our clients are not always the most technically inclined, and more than once I've had to restore a revision to fix something that "mysteriously vanished" (nearly always user error).

In the ACF Forum thread you commented on the serializing method and circumventing update_meta, but you didn't remark on the idea of marking fields as "dirty" if they change. I know that just reduces the impact of the problem without really solving it, but it seems like it might be an easier optimization than recreating the entire save process, particularly if revisions are part of the problem.

Is this worth pursuing or is there something you know of that would make it unfeasible?

**Comment by Hube2** [comment]
I've just been thinking about it and a thought struck me, and this thought was partially inspired by the idea you posted about saving repeaters and flex fields as serialized data.

Most of the problems happen because of repeaters, flex fields and clones. The nested and recursive nature of them. If we eliminate the saving of sub field entirely and did serialize them that could eliminate the problem. Some people want these fields to be query-able, but I consider that an edge case.  The amount of effort required to query for a value in a repeater sub field is ridiculous. I've posted many times that in these cases that I create filters to save the contents of these fields into other standard WP meta fields to make them more easily searched.

The problem to overcome here is how to insert the values when one of the ACF function, filters and actions are called. There would be a lot of these to deal with and I'm not sure it could be done cleanly.

**Comment by jakefinley** [comment]
That would be acceptable to me, I've only once ever felt the need to to query a repeater sub field and if I'd wanted I could have done it a different way easily.

I assume that ACF only fetches the fields from the database once, and then caches them. Wouldn't you just need to get the unserialized version of the data into ACF's internal array, and then from there ACF could handle the data normally? Failing that, potentially it might even be simpler to do this from the wordpress side of things and filter get_metadata, feeding ACF its expected results after prepossessing the request.

**Comment by Hube2** [comment]
ACF uses the WP meta cache, so the trick would be to insert them into that. I don't even know if that's possible.

Of perhaps to return a value from the array on the acf/load_value filter. You'd also need to all the acf/update_value filters on every field before saving the value, and the update value filters would need to run even if you didn't use acf filters for the load value and somehow worked with the cache. 

I've actually got another thought that I think would be infinitely less complicated to do that I'm looking into.

**Comment by Hube2** [comment]
Here's another idea in a completely different direction.

I don't know about anyone else, but my main concern is admin pages that time out. I don't really care if the editor has to wait a few more seconds after saving a page. The only time I care about load times is on the front end of the site and ACF performs well there. The long load time is only a problem when the page times out because I have to hear about it every time it does and they forgot what I told them the last time.

Here is how this plan would work.

1) add and acf/save_post action with s ridiculously low priority, like - 99999. Something that no one should ever think to use something lower than.

2) Set PHP to not stop processing, I've done this in other places where I knew the process would more than likely time out.

3) Set a value in the database, a flag that indicates that the post is being updated.

4) Send a redirect to a middle page, or even output that middle page.

5) The user sees a page that says something like "Your request is being processed" This page includes a JS function that runs on a timeout that does an AJAX request to see if the processing is done.

6) When the process is done the user is redirected back to the post edit page.

7) Meanwhile, after 4 is done, the PHP process continues and everything is saved the way it should be.

8) another acf/save_post action with a ridiculously high priory unsets the flag in the database which allow the user to be redirected back to the post editor page.

Like I said, a completely different direction. It gets the job done and is infinitely less complicated to achieve. This code would take no work whatsoever to maintain once it was put in place.

**Comment by jakefinley** [comment]
That definitely wouldn't serve my purposes. For me it is not about timeouts (though that has happened), but that I've actually had customers complain about how slow it is to update pages. This also effects previews, though not as badly (probably because of revisions), and generally just slows down content editing.

The one situation where this particular functionality would have help is when I was working on a Rackspace Cloud Site, in that situation I couldn't control their load balancer which would cut any connection after 30s. Pages would update, but they'd nearly always time out and I couldn't do anything about it.

**Comment by Hube2** [comment]
I've created another repo for the last train of thought, it's a different direction completely. I'm not going to abandon this, but this is a huge complicated project no matter what way it goes. I have some time right now to continue researching it, but that time could go away at any moment. Whereas I can probably get the other thing working in an afternoon and for me a temporary bandage is better than dying while I wait for the doctor.

**Comment by ctaity** [comment]
Can implement save the metadata in mongodb, 1 write 1 read, for all custom fields?
We are developers, if some one can guide us, we can do it.
